File: Sources/BNKKit/Kubeconfig.swift
import Foundation
import Yams

/// A kubeconfig, reduced to the parts Field can actually act on.
///
/// Deliberately not a faithful model of the format. bnkscope's own normalizer
/// exists because kubeconfigs in the wild carry auth shapes that need a binary
/// on PATH; on iOS there is no PATH, so those contexts are recorded with the
/// reason they cannot be used rather than half-loaded and failed later.
public struct Kubeconfig: Sendable {
    public let contexts: [Context]

    public struct Context: Sendable {
        public let name: String
        public let clusterName: String
        public let server: URL
        public let caPEM: Data?
        /// The name the server's certificate is checked against, when that is
        /// not the address dialled.
        ///
        /// A cluster reached through a forward — a lab apiserver bound to
        /// loopback, published on another address — presents a certificate for
        /// the name it knows itself by, not the one you connected to. Ignoring
        /// this field means such a cluster can only be used by turning
        /// verification off altogether, which is a much bigger hammer.
        public let tlsServerName: String?
        public let insecureSkipTLSVerify: Bool
        public let namespace: String?
        public let auth: Auth
    }

    /// What Field can present to an apiserver. `.unsupported` is a first-class
    /// case, not an error: the context is listed with its reason.
    public enum Auth: Sendable, Equatable {
        case clientCertificate(certPEM: Data, keyPEM: Data)
        case bearerToken(String)
        case unsupported(reason: String)
    }

    public enum ParseError: Error, CustomStringConvertible {
        case notAMapping
        case noContexts
        case missingCluster(String)
        case badServer(String)

        public var description: String {
            switch self {
            case .notAMapping:        return "kubeconfig is not a YAML mapping"
            case .noContexts:         return "kubeconfig lists no contexts"
            case .missingCluster(let n): return "context references cluster \"\(n)\", which is not defined"
            case .badServer(let s):   return "cluster server \"\(s)\" is not a URL"
            }
        }
    }

    public init(yaml: String) throws {
        guard let root = try Yams.load(yaml: yaml) as? [String: Any] else {
            throw ParseError.notAMapping
        }
        let clusters = Self.byName(root["clusters"], key: "cluster")
        let users    = Self.byName(root["users"], key: "user")
        let raw      = root["contexts"] as? [[String: Any]] ?? []
        guard !raw.isEmpty else { throw ParseError.noContexts }

        var out: [Context] = []
        for entry in raw {
            guard let name = entry["name"] as? String,
                  let ctx  = entry["context"] as? [String: Any],
                  let clusterName = ctx["cluster"] as? String else { continue }
            guard let cluster = clusters[clusterName] else {
                throw ParseError.missingCluster(clusterName)
            }
            guard let serverString = cluster["server"] as? String,
                  let server = URL(string: serverString) else {
                throw ParseError.badServer((cluster["server"] as? String) ?? "")
            }
            let user = (ctx["user"] as? String).flatMap { users[$0] } ?? [:]
            out.append(Context(
                name: name,
                clusterName: clusterName,
                server: server,
                caPEM: Self.pemOrFile(cluster, dataKey: "certificate-authority-data", fileKey: "certificate-authority"),
                tlsServerName: cluster["tls-server-name"] as? String,
                insecureSkipTLSVerify: (cluster["insecure-skip-tls-verify"] as? Bool) ?? false,
                namespace: ctx["namespace"] as? String,
                auth: Self.auth(from: user)))
        }
        self.contexts = out
    }

    public init(contentsOf url: URL) throws {
        try self.init(yaml: String(contentsOf: url, encoding: .utf8))
    }

    public func context(named name: String) -> Context? {
        contexts.first { $0.name == name }
    }

    /// One self-contained kubeconfig per context.
    ///
    /// A file with three contexts in it is three clusters, and after import it
    /// should be three independent things — removing one has no business
    /// touching the others. Splitting on the way in makes that true by
    /// construction, rather than by remembering which contexts were removed and
    /// filtering them out on every load.
    public static func split(yaml: String) throws -> [(name: String, yaml: String)] {
        guard let root = try Yams.load(yaml: yaml) as? [String: Any] else {
            throw ParseError.notAMapping
        }
        let clusters = byName(root["clusters"], key: "cluster")
        let users = byName(root["users"], key: "user")
        var out: [(String, String)] = []

        for entry in root["contexts"] as? [[String: Any]] ?? [] {
            guard let name = entry["name"] as? String,
                  let ctx = entry["context"] as? [String: Any],
                  let clusterName = ctx["cluster"] as? String,
                  let cluster = clusters[clusterName] else { continue }
            let userName = ctx["user"] as? String
            var document: [String: Any] = [
                "apiVersion": "v1",
                "kind": "Config",
                "clusters": [["name": clusterName, "cluster": cluster]],
                "contexts": [["name": name, "context": ctx]],
                "current-context": name,
            ]
            if let userName, let user = users[userName] {
                document["users"] = [["name": userName, "user": user]]
            }
            guard let text = try? Yams.dump(object: document) else { continue }
            out.append((name, text))
        }
        return out
    }

    // MARK: - Auth

    static func auth(from user: [String: Any]) -> Auth {
        // exec: and auth-provider: both mean "run this program", which this app
        // does not do — on any platform. The rule is the app's, not the OS's,
        // and the wording says so: this file is shared with the Mac build and
        // the Linux CLI, and "iOS runs no binaries" was being shown on a Mac.
        // Name the binary in the reason — the user has to replace it with a
        // certificate or a token, and needs to know which one is in the way.
        if let exec = user["exec"] as? [String: Any] {
            let cmd = (exec["command"] as? String) ?? "an external command"
            return .unsupported(reason: "needs `\(cmd)`, and this app runs no binaries — supply a client certificate or a bearer token instead")
        }
        if let provider = user["auth-provider"] as? [String: Any] {
            let name = (provider["name"] as? String) ?? "an auth provider"
            return .unsupported(reason: "uses the `\(name)` auth provider, which needs a helper binary — supply a client certificate or a bearer token instead")
        }
        if let token = user["token"] as? String, !token.isEmpty {
            return .bearerToken(token)
        }
        if let cert = pemOrFile(user, dataKey: "client-certificate-data", fileKey: "client-certificate"),
           let key  = pemOrFile(user, dataKey: "client-key-data",         fileKey: "client-key") {
            return .clientCertificate(certPEM: cert, keyPEM: key)
        }
        if user["username"] != nil {
            return .unsupported(reason: "uses HTTP basic auth, which Kubernetes removed in 1.19")
        }
        return .unsupported(reason: "carries no credentials Field can present")
    }

    // MARK: - Helpers

    /// `*-data` is base64 inline; the plain key is a path on the machine that
    /// wrote the file. On iOS that path does not exist, so a file reference is
    /// read when it happens to resolve (macOS, tests) and otherwise dropped —
    /// the caller reports the context as unusable rather than guessing.
    static func pemOrFile(_ m: [String: Any], dataKey: String, fileKey: String) -> Data? {
        if let b64 = m[dataKey] as? String, let d = Data(base64Encoded: b64, options: .ignoreUnknownCharacters) {
            return d
        }
        if let path = m[fileKey] as? String {
            return try? Data(contentsOf: URL(fileURLWithPath: path))
        }
        return nil
    }

    static func byName(_ any: Any?, key: String) -> [String: [String: Any]] {
        guard let list = any as? [[String: Any]] else { return [:] }
        var out: [String: [String: Any]] = [:]
        for e in list {
            guard let name = e["name"] as? String else { continue }
            out[name] = (e[key] as? [String: Any]) ?? [:]
        }
        return out
    }
}

File: dart/bnk_kit/lib/src/kubeconfig.dart
import 'dart:convert';
import 'dart:io';
import 'dart:typed_data';

import 'package:yaml/yaml.dart';

import 'yaml_emit.dart';

/// What Field can present to an apiserver. [UnsupportedAuth] is a first-class
/// case, not an error: the context is listed with its reason.
sealed class KubeAuth {
  const KubeAuth();
}

class ClientCertificateAuth extends KubeAuth {
  final Uint8List certPEM;
  final Uint8List keyPEM;
  const ClientCertificateAuth(this.certPEM, this.keyPEM);

  @override
  bool operator ==(Object other) =>
      other is ClientCertificateAuth &&
      _bytesEqual(other.certPEM, certPEM) &&
      _bytesEqual(other.keyPEM, keyPEM);

  @override
  int get hashCode => Object.hash(certPEM.length, keyPEM.length);
}

class BearerTokenAuth extends KubeAuth {
  final String token;
  const BearerTokenAuth(this.token);

  @override
  bool operator ==(Object other) =>
      other is BearerTokenAuth && other.token == token;

  @override
  int get hashCode => token.hashCode;
}

class UnsupportedAuth extends KubeAuth {
  final String reason;
  const UnsupportedAuth(this.reason);

  @override
  bool operator ==(Object other) =>
      other is UnsupportedAuth && other.reason == reason;

  @override
  int get hashCode => reason.hashCode;
}

bool _bytesEqual(List<int> a, List<int> b) {
  if (a.length != b.length) return false;
  for (var i = 0; i < a.length; i++) {
    if (a[i] != b[i]) return false;
  }
  return true;
}

class KubeContext {
  final String name;
  final String clusterName;
  final Uri server;
  final Uint8List? caPEM;

  /// The name the server's certificate is checked against, when that is not
  /// the address dialled.
  ///
  /// A cluster reached through a forward, a lab apiserver bound to loopback
  /// and published on another address, presents a certificate for the name it
  /// knows itself by, not the one you connected to. Ignoring this field means
  /// such a cluster can only be used by turning verification off altogether,
  /// which is a much bigger hammer.
  final String? tlsServerName;
  final bool insecureSkipTLSVerify;
  final String? namespace;
  final KubeAuth auth;

  const KubeContext({
    required this.name,
    required this.clusterName,
    required this.server,
    required this.caPEM,
    required this.tlsServerName,
    required this.insecureSkipTLSVerify,
    required this.namespace,
    required this.auth,
  });
}

class KubeconfigException implements Exception {
  final String message;
  const KubeconfigException(this.message);

  const KubeconfigException.notAMapping()
      : message = 'kubeconfig is not a YAML mapping';
  const KubeconfigException.noContexts()
      : message = 'kubeconfig lists no contexts';
  KubeconfigException.missingCluster(String name)
      : message = 'context references cluster "$name", which is not defined';
  KubeconfigException.badServer(String server)
      : message = 'cluster server "$server" is not a URL';

  @override
  String toString() => message;
}

/// A kubeconfig, reduced to the parts Field can actually act on.
///
/// Deliberately not a faithful model of the format. Kubeconfigs in the wild
/// carry auth shapes that need a binary on PATH; this app runs none, so those
/// contexts are recorded with the reason they cannot be used rather than
/// half-loaded and failed later.
class Kubeconfig {
  final List<KubeContext> contexts;

  const Kubeconfig._(this.contexts);

  factory Kubeconfig.parse(String yaml) {
    final root = _root(yaml);
    final clusters = _byName(root['clusters'], 'cluster');
    final users = _byName(root['users'], 'user');
    final raw = _mapList(root['contexts']);
    if (raw.isEmpty) throw const KubeconfigException.noContexts();

    final out = <KubeContext>[];
    for (final entry in raw) {
      final name = entry['name'];
      final ctx = _map(entry['context']);
      final clusterName = ctx?['cluster'];
      if (name is! String || ctx == null || clusterName is! String) continue;
      final cluster = clusters[clusterName];
      if (cluster == null) throw KubeconfigException.missingCluster(clusterName);
      final serverString = cluster['server'];
      final server = serverString is String ? Uri.tryParse(serverString) : null;
      if (server == null || !server.hasScheme || server.host.isEmpty) {
        throw KubeconfigException.badServer(
            serverString is String ? serverString : '');
      }
      final userName = ctx['user'];
      final user =
          (userName is String ? users[userName] : null) ?? <String, dynamic>{};
      out.add(KubeContext(
        name: name,
        clusterName: clusterName,
        server: server,
        caPEM: pemOrFile(cluster, 'certificate-authority-data',
            'certificate-authority'),
        tlsServerName: cluster['tls-server-name'] as String?,
        insecureSkipTLSVerify:
            cluster['insecure-skip-tls-verify'] as bool? ?? false,
        namespace: ctx['namespace'] as String?,
        auth: authFrom(user),
      ));
    }
    return Kubeconfig._(out);
  }

  static Future<Kubeconfig> load(String path) async =>
      Kubeconfig.parse(await File(path).readAsString());

  factory Kubeconfig.loadSync(String path) =>
      Kubeconfig.parse(File(path).readAsStringSync());

  KubeContext? context(String name) {
    for (final c in contexts) {
      if (c.name == name) return c;
    }
    return null;
  }

  /// One self-contained kubeconfig per context.
  ///
  /// A file with three contexts in it is three clusters, and after import it
  /// should be three independent things: removing one has no business touching
  /// the others. Splitting on the way in makes that true by construction,
  /// rather than by remembering which contexts were removed and filtering them
  /// out on every load.
  static List<({String name, String yaml})> split(String yaml) {
    final root = _root(yaml);
    final clusters = _byName(root['clusters'], 'cluster');
    final users = _byName(root['users'], 'user');
    final out = <({String name, String yaml})>[];

    for (final entry in _mapList(root['contexts'])) {
      final name = entry['name'];
      final ctx = _map(entry['context']);
      final clusterName = ctx?['cluster'];
      if (name is! String || ctx == null || clusterName is! String) continue;
      final cluster = clusters[clusterName];
      if (cluster == null) continue;
      final userName = ctx['user'];
      final document = <String, dynamic>{
        'apiVersion': 'v1',
        'kind': 'Config',
        'clusters': [
          {'name': clusterName, 'cluster': cluster}
        ],
        'contexts': [
          {'name': name, 'context': ctx}
        ],
        'current-context': name,
      };
      final user = userName is String ? users[userName] : null;
      if (user != null) {
        document['users'] = [
          {'name': userName, 'user': user}
        ];
      }
      out.add((name: name, yaml: emitYaml(document)));
    }
    return out;
  }

  // Auth

  static KubeAuth authFrom(Map<String, dynamic> user) {
    // exec: and auth-provider: both mean "run this program", which this app
    // does not do, on any platform. Name the binary in the reason: the user
    // has to replace it with a certificate or a token, and needs to know
    // which one is in the way.
    final exec = _map(user['exec']);
    if (exec != null) {
      final cmd = exec['command'] as String? ?? 'an external command';
      return UnsupportedAuth(
          'needs `$cmd`, and this app runs no binaries — supply a client certificate or a bearer token instead');
    }
    final provider = _map(user['auth-provider']);
    if (provider != null) {
      final name = provider['name'] as String? ?? 'an auth provider';
      return UnsupportedAuth(
          'uses the `$name` auth provider, which needs a helper binary — supply a client certificate or a bearer token instead');
    }
    final token = user['token'];
    if (token is String && token.isNotEmpty) return BearerTokenAuth(token);
    final cert =
        pemOrFile(user, 'client-certificate-data', 'client-certificate');
    final key = pemOrFile(user, 'client-key-data', 'client-key');
    if (cert != null && key != null) return ClientCertificateAuth(cert, key);
    if (user['username'] != null) {
      return const UnsupportedAuth(
          'uses HTTP basic auth, which Kubernetes removed in 1.19');
    }
    return const UnsupportedAuth('carries no credentials Field can present');
  }

  // Helpers

  static final _notBase64 = RegExp(r'[^A-Za-z0-9+/=]');

  /// `*-data` is base64 inline; the plain key is a path on the machine that
  /// wrote the file. On a tablet that path does not exist, so a file reference
  /// is read when it happens to resolve (a desktop, tests) and otherwise
  /// dropped; the caller reports the context as unusable rather than guessing.
  static Uint8List? pemOrFile(
      Map<String, dynamic> m, String dataKey, String fileKey) {
    final b64 = m[dataKey];
    if (b64 is String) {
      try {
        return base64.decode(base64.normalize(b64.replaceAll(_notBase64, '')));
      } on FormatException {
        // fall through to the file form
      }
    }
    final path = m[fileKey];
    if (path is String) {
      try {
        return File(path).readAsBytesSync();
      } on IOException {
        return null;
      }
    }
    return null;
  }

  static Map<String, dynamic> _root(String yaml) {
    final Object? loaded;
    try {
      loaded = loadYaml(yaml);
    } on YamlException catch (e) {
      throw KubeconfigException('kubeconfig is not valid YAML: ${e.message}');
    }
    final root = _plain(loaded);
    if (root is! Map<String, dynamic>) {
      throw const KubeconfigException.notAMapping();
    }
    return root;
  }

  static Map<String, Map<String, dynamic>> _byName(Object? any, String key) {
    final out = <String, Map<String, dynamic>>{};
    for (final e in _mapList(any)) {
      final name = e['name'];
      if (name is! String) continue;
      out[name] = _map(e[key]) ?? <String, dynamic>{};
    }
    return out;
  }

  static List<Map<String, dynamic>> _mapList(Object? any) {
    if (any is! List) return const [];
    return [
      for (final e in any)
        if (e is Map<String, dynamic>) e
    ];
  }

  static Map<String, dynamic>? _map(Object? any) =>
      any is Map<String, dynamic> ? any : null;

  /// YAML nodes to plain Dart values, with string keys throughout.
  static Object? _plain(Object? node) {
    if (node is Map) {
      return <String, dynamic>{
        for (final e in node.entries) e.key.toString(): _plain(e.value)
      };
    }
    if (node is List) return [for (final e in node) _plain(e)];
    return node;
  }
}
